From d560bc45ac5d02fa3fab7ffb46c894063210cb56 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:07:21 -0400
Subject: [usb] usb_wwan: fix handling of missing bulk endpoints

Message-id: <1410887463-20674-370-git-send-email-dzickus@redhat.com>
Patchwork-id: 95587
O-Subject: [RHEL7 PATCH 369/591] USB: usb_wwan: fix handling of missing bulk endpoints
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit bd73bd8831696f189a479a0712ae95208e513d7e
Author: Johan Hovold <jhovold@gmail.com>
Date:   Thu Apr 3 13:06:46 2014 +0200

    USB: usb_wwan: fix handling of missing bulk endpoints

    Fix regression introduced by commit 8e493ca1767d ("USB: usb_wwan: fix
    bulk-urb allocation") by making sure to require both bulk-in and out
    endpoints during port probe.

    The original option driver (which usb_wwan is based on) was written
    under the assumption that either endpoint could be missing, but
    evidently this cannot have been tested properly. Specifically, it would
    handle opening a device without bulk-in (but would blow up during resume
    which was implemented later), but not a missing bulk-out in write()
    (although it is handled in some places such as write_room()).

    Fortunately (?), the driver also got the test for missing endpoints
    wrong so the urbs were in fact always allocated, although they would be
    initialised using the wrong endpoint address (0) and any submission of
    such an urb would fail.

    The commit mentioned above fixed the test for missing endpoints but
    thereby exposed the other bugs which would now generate null-pointer
    exceptions rather than failed urb submissions.

    The regression was introduced in v3.7, but the offending commit was also
    marked for stable.

    Reported-by: Rafal Milecki <zajec5@gmail.com>
    Cc: stable <stable@vger.kernel.org>
    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Tested-by: Rafal Milecki <zajec5@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/usb_wwan.c b/drivers/usb/serial/usb_wwan.c
index 640fe01..b078440 100644
--- a/drivers/usb/serial/usb_wwan.c
+++ b/drivers/usb/serial/usb_wwan.c
@@ -466,6 +466,9 @@ int usb_wwan_port_probe(struct usb_serial_port *port)
  int err;
  int i;
 
+ if (!port->bulk_in_size || !port->bulk_out_size)
+  return -ENODEV;
+
  portdata = kzalloc(sizeof(*portdata), GFP_KERNEL);
  if (!portdata)
   return -ENOMEM;
@@ -473,9 +476,6 @@ int usb_wwan_port_probe(struct usb_serial_port *port)
  init_usb_anchor(&portdata->delayed);
 
  for (i = 0; i < N_IN_URB; i++) {
-  if (!port->bulk_in_size)
-   break;
-
   buffer = (u8 *)__get_free_page(GFP_KERNEL);
   if (!buffer)
    goto bail_out_error;
@@ -489,9 +489,6 @@ int usb_wwan_port_probe(struct usb_serial_port *port)
  }
 
  for (i = 0; i < N_OUT_URB; i++) {
-  if (!port->bulk_out_size)
-   break;
-
   buffer = kmalloc(OUT_BUFLEN, GFP_KERNEL);
   if (!buffer)
    goto bail_out_error2;
-- 
1.7.1